File: backend/src/datasource/dbdao/content.rs
#![allow(dead_code)]
use super::dao::DBDao;
use anyhow::Result;
use serde::{Deserialize, Serialize};
use sqlx::Row;

#[derive(Deserialize, Serialize, Debug, Clone)]
pub struct ContentRecord {
    pub id: i64,
    pub title: String,
    pub content: String,
    pub content_type: Option<String>,
    pub status: Option<String>,
    pub tags: Option<String>,
    pub created_at: Option<String>,
    pub published_at: Option<String>,
    pub updated_at: Option<String>,
}

impl DBDao {
    /// Get content by ID
    pub async fn get_content(&self, id: i64) -> Result<ContentRecord> {
        let row = sqlx::query(
            "SELECT id, title, content, content_type, status, tags, created_at, published_at, updated_at
             FROM content WHERE id = $1")
        .bind(id)
        .fetch_one(&self.db)
        .await?;

        let content = ContentRecord {
            id: row.get("id"),
            title: row.get("title"),
            content: row.get("content"),
            content_type: row.get("content_type"),
            status: row.get("status"),
            tags: row.get("tags"),
            created_at: row.get("created_at"),
            published_at: row.get("published_at"),
            updated_at: row.get("updated_at"),
        };

        Ok(content)
    }

    /// Search content with pagination
    pub async fn search_contents(
        &self,
        query: Option<&str>,
        _content_type: Option<String>,
        _status: Option<String>,
        _tag: Option<String>,
        page: i64,
        limit: i64,
    ) -> Result<Vec<ContentRecord>> {
        let offset = (page - 1) * limit;
        let rows = sqlx::query(
            "SELECT id, title, content, content_type, status, tags, created_at, published_at, updated_at
             FROM content
             WHERE $1 IS NULL OR content LIKE $1
             OFFSET $2 LIMIT $3")
        .bind(query.unwrap_or("%%"))
        .bind(offset)
        .bind(limit)
        .fetch_all(&self.db)
        .await?;

        let contents = rows
            .into_iter()
            .map(|row| ContentRecord {
                id: row.get("id"),
                title: row.get("title"),
                content: row.get("content"),
                content_type: row.get("content_type"),
                status: row.get("status"),
                tags: row.get("tags"),
                created_at: row.get("created_at"),
                published_at: row.get("published_at"),
                updated_at: row.get("updated_at"),
            })
            .collect::<Vec<_>>();

        Ok(contents)
    }

    /// Add content
    pub async fn add_content(
        &self,
        title: &str,
        content: &str,
        content_type: Option<String>,
        tags: Option<String>,
    ) -> Result<i64> {
        let row = sqlx::query(
            "INSERT INTO content (title, content, content_type, tags)
             VALUES ($1, $2, $3, $4)
             RETURNING id",
        )
        .bind(title)
        .bind(content)
        .bind(content_type)
        .bind(tags)
        .fetch_one(&self.db)
        .await?;

        Ok(row.get("id"))
    }

    /// Update content
    pub async fn update_content(
        &self,
        id: i64,
        title: Option<String>,
        content: Option<String>,
        content_type: Option<String>,
        tags: Option<String>,
    ) -> Result<i64> {
        let row = sqlx::query(
            "UPDATE content
             SET title = COALESCE($1, title),
                 content = COALESCE($2, content),
                 content_type = COALESCE($3, content_type),
                 tags = COALESCE($4, tags),
                 updated_at = NOW()
             WHERE id = $5
             RETURNING id",
        )
        .bind(title)
        .bind(content)
        .bind(content_type)
        .bind(tags)
        .bind(id)
        .fetch_one(&self.db)
        .await?;

        Ok(row.get("id"))
    }

    /// Delete content
    pub async fn delete_content(&self, id: i64) -> Result<i64> {
        let row = sqlx::query("DELETE FROM content WHERE id = $1 RETURNING id")
            .bind(id)
            .fetch_one(&self.db)
            .await?;

        Ok(row.get("id"))
    }

    /// Create content version
    pub async fn create_content_version(
        &self,
        content_id: i64,
        content: &str,
        created_by: Option<String>,
    ) -> Result<i64> {
        let row = sqlx::query(
            "INSERT INTO content_versions (content_id, content, created_by)
             VALUES ($1, $2, $3)
             RETURNING id",
        )
        .bind(content_id)
        .bind(content)
        .bind(created_by)
        .fetch_one(&self.db)
        .await?;

        Ok(row.get("id"))
    }

    /// Publish content
    pub async fn publish_content(
        &self,
        id: i64,
        published_at: Option<String>,
        published_by: Option<String>,
    ) -> Result<i64> {
        let row = sqlx::query(
            "UPDATE content
             SET status = 'published',
                 published_at = COALESCE($1, published_at),
                 published_by = COALESCE($2, published_by)
             WHERE id = $3
             RETURNING id",
        )
        .bind(published_at)
        .bind(published_by)
        .bind(id)
        .fetch_one(&self.db)
        .await?;

        Ok(row.get("id"))
    }
}

File: backend/datasource/dbdao/content.go
package dbdao

import (
	"context"
	"encoding/json"
	"errors"
	"time"

	"gorm.io/gorm"
)

// ContentModel 数据库模型
type ContentModel struct {
	ID          int64     `gorm:"primarykey;column:id"`
	Title       string    `gorm:"column:title;type:varchar(255);not null"`
	Body        string    `gorm:"column:body;type:longtext"`
	ContentType string    `gorm:"column:content_type;type:varchar(100);not null"`
	Status      string    `gorm:"column:status;type:varchar(50)"`
	Tags        string    `gorm:"column:tags;type:json"`
	CreatedAt   time.Time `gorm:"column:created_at;autoCreateTime:milli"`
	UpdatedAt   time.Time `gorm:"column:updated_at;autoUpdateTime:milli"`
	PublishedAt *time.Time `gorm:"column:published_at"`
}

// TableName 指定表名
func (ContentModel) TableName() string {
	return "contents"
}

// Content DAO层返回的内容数据结构
type Content struct {
	ID          int64
	Title       string
	Body        string
	ContentType string
	Status      string
	Tags        []string
	CreatedAt   time.Time
	UpdatedAt   time.Time
	PublishedAt *time.Time
}

// GetContent 获取内容
func (d *DB) GetContent(ctx context.Context, id int64) (*Content, error) {
	var model ContentModel
	result := d.DB().WithContext(ctx).Where("id = ?", id).First(&model)
	if result.Error != nil {
		if result.Error == gorm.ErrRecordNotFound {
			return nil, errors.New("content not found")
		}
		return nil, result.Error
	}

	var tags []string
	if len(model.Tags) > 0 {
		if err := json.Unmarshal([]byte(model.Tags), &tags); err != nil {
			tags = []string{}
		}
	}

	return &Content{
		ID:          model.ID,
		Title:       model.Title,
		Body:        model.Body,
		ContentType: model.ContentType,
		Status:      model.Status,
		Tags:        tags,
		CreatedAt:   model.CreatedAt,
		UpdatedAt:   model.UpdatedAt,
		PublishedAt: model.PublishedAt,
	}, nil
}

// SearchContents 搜索内容
func (d *DB) SearchContents(
	ctx context.Context,
	query, contentType, status, tag string,
	page, limit int64,
) ([]*Content, int64, error) {
	var models []ContentModel
	q := d.DB().WithContext(ctx)

	if query != "" {
		q = q.Where("title LIKE ? OR body LIKE ?", "%"+query+"%", "%"+query+"%")
	}
	if contentType != "" {
		q = q.Where("content_type = ?", contentType)
	}
	if status != "" {
		q = q.Where("status = ?", status)
	}
	if tag != "" {
		q = q.Where("JSON_CONTAINS(tags, ?)", `"`+tag+`"`)
	}

	var total int64
	q.Model(&ContentModel{}).Count(&total)

	result := q.Offset(int((page - 1) * limit)).Limit(int(limit)).Find(&models)
	if result.Error != nil {
		return nil, 0, result.Error
	}

	contents := make([]*Content, len(models))
	for i, model := range models {
		var tags []string
		if len(model.Tags) > 0 {
			if err := json.Unmarshal([]byte(model.Tags), &tags); err != nil {
				tags = []string{}
			}
		}

		contents[i] = &Content{
			ID:          model.ID,
			Title:       model.Title,
			Body:        model.Body,
			ContentType: model.ContentType,
			Status:      model.Status,
			Tags:        tags,
			CreatedAt:   model.CreatedAt,
			UpdatedAt:   model.UpdatedAt,
			PublishedAt: model.PublishedAt,
		}
	}

	return contents, total, nil
}

// CreateContent 创建内容
func (d *DB) CreateContent(ctx context.Context, content *Content) (*Content, error) {
	tagsJSON, err := json.Marshal(content.Tags)
	if err != nil {
		tagsJSON = []byte("[]")
	}

	model := &ContentModel{
		Title:       content.Title,
		Body:        content.Body,
		ContentType: content.ContentType,
		Status:      content.Status,
		Tags:        string(tagsJSON),
	}

	result := d.DB().WithContext(ctx).Create(model)
	if result.Error != nil {
		return nil, result.Error
	}

	return &Content{
		ID:          model.ID,
		Title:       model.Title,
		Body:        model.Body,
		ContentType: model.ContentType,
		Status:      model.Status,
		Tags:        content.Tags,
		CreatedAt:   model.CreatedAt,
		UpdatedAt:   model.UpdatedAt,
		PublishedAt: model.PublishedAt,
	}, nil
}

// UpdateContent 更新内容
func (d *DB) UpdateContent(ctx context.Context, content *Content) (*Content, error) {
	updates := map[string]interface{}{}

	if content.Title != "" {
		updates["title"] = content.Title
	}
	if content.Body != "" {
		updates["body"] = content.Body
	}
	if content.ContentType != "" {
		updates["content_type"] = content.ContentType
	}
	if content.Status != "" {
		updates["status"] = content.Status
	}
	if len(content.Tags) > 0 {
		tagsJSON, _ := json.Marshal(content.Tags)
		updates["tags"] = string(tagsJSON)
	}

	if len(updates) == 0 {
		return nil, errors.New("no fields to update")
	}

	result := d.DB().WithContext(ctx).Model(&ContentModel{}).Where("id = ?", content.ID).Updates(updates)
	if result.Error != nil {
		return nil, result.Error
	}

	if result.RowsAffected == 0 {
		return nil, errors.New("content not found")
	}

	return d.GetContent(ctx, content.ID)
}

// DeleteContent 删除内容
func (d *DB) DeleteContent(ctx context.Context, id int64) error {
	result := d.DB().WithContext(ctx).Where("id = ?", id).Delete(&ContentModel{})
	if result.Error != nil {
		return result.Error
	}

	if result.RowsAffected == 0 {
		return errors.New("content not found")
	}

	return nil
}
